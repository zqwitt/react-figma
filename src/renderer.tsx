import * as React from 'react';
import * as renderers from './renderers';
// todo replace with webpack aliasing, install @types/react-reconciler
import * as createReconciler from './realm-adopted/react-reconciler';

import { GroupsProcessor } from './renderers/group/groupsProcessor';
import { PREGROUP_NODE_TYPE } from './renderers/group/pregroupNode';

const isReactFigmaNode = child => child.getPluginData && child.getPluginData('isReactFigmaNode');

const appendToContainerFactory = groupsProcessor => (parentNode, childNode) => {
    if (!childNode || !parentNode) {
        return;
    }

    if (childNode.type === PREGROUP_NODE_TYPE) {
        if (childNode.groupNode) {
            parentNode.appendChild(childNode.groupNode);
        } else {
            groupsProcessor.scheduleGroup(parentNode, childNode);
        }
    } else {
        parentNode.appendChild(childNode);
    }
};

const insertToContainer = (parentNode, newChildNode, beforeChildNode) => {
    if (!parentNode || !newChildNode || !beforeChildNode) {
        return;
    }
    const beforeChildIndex = parentNode.children.indexOf(beforeChildNode);
    parentNode.insertChild(beforeChildIndex, newChildNode);
};

const remove = childNode => {
    if (!childNode || childNode.removed) {
        return;
    }
    childNode.remove();
};

const renderInstance = (type, node, props) => {
    const instance = renderers[type](node)(props);
    if (!node) {
        instance.setPluginData('isReactFigmaNode', 'true');
    }
    return instance;
};

const getFirstChild = parentInstance => {
    if (parentInstance.children && parentInstance.children.length > 0) {
        return parentInstance.children.find(isReactFigmaNode);
    }
};

const getNextChildren = instance => {
    if (!instance || !instance.parent) {
        return;
    }
    const parent = instance.parent;
    console.log('getNextHydratableSibling:children', parent.children);
    const instanceIndex = parent.children.indexOf(instance);
    console.log('getNextHydratableSibling:instanceIndex', instanceIndex);
    return parent.children.slice(instanceIndex + 1).find(isReactFigmaNode);
};

export const render = async (jsx: any, rootNode) => {
    const groupsProcessor = new GroupsProcessor();
    const appendToContainer = appendToContainerFactory(groupsProcessor);

    const HostConfig = {
        now: Date.now,
        getRootHostContext: (...args) => {
            console.log('getRootHostContext', ...args);
            return true;
        },
        prepareForCommit: (...args) => {
            console.log('prepareForCommit', ...args);
        },
        resetAfterCommit: (...args) => {
            console.log('resetAfterCommit', ...args);
        },
        getChildHostContext: (...args) => {
            console.log('getChildHostContext', ...args);
            return true;
        },
        shouldSetTextContent: () => false,
        getPublicInstance: instance => {
            console.log('getPublicInstance', instance);
            return instance;
        },
        createInstance: (type, props, ...other) => {
            console.log('createInstance', type, props, ...other);
            return renderInstance(type, null, props);
        },
        createTextInstance: text => {
            console.log('createTextInstance', text);
        },
        resetTextContent: node => {
            console.log('resetTextContent', node);
        },
        // Append root node to a container
        appendInitialChild: (parentNode, childNode) => {
            console.log('appendInitialChild', parentNode, childNode);
            appendToContainer(parentNode, childNode);
        },
        appendChild: (parentNode, childNode) => {
            console.log('appendChild', parentNode, childNode);
            appendToContainer(parentNode, childNode);
        },
        insertBefore: (parentNode, newChildNode, beforeChildNode) => {
            console.log('insertBefore', parentNode, newChildNode, beforeChildNode);
            insertToContainer(parentNode, newChildNode, beforeChildNode);
        },
        finalizeInitialChildren: (element, type, ...args) => {
            console.log('finalizeInitialChildren', type, ...args);

            return type === 'page';
        },
        supportsMutation: true,
        supportsHydration: true,
        appendChildToContainer: (parentNode, childNode) => {
            console.log('appendChildToContainer', parentNode, childNode);
            appendToContainer(parentNode, childNode);
        },
        insertInContainerBefore: (parentNode, newChildNode, beforeChildNode) => {
            console.log('insertInContainerBefore', parentNode, newChildNode, beforeChildNode);
        },
        removeChildFromContainer: (parentNode, childNode) => {
            console.log('removeChildFromContainer', parentNode, childNode);
        },
        prepareUpdate: (...args) => {
            console.log('prepareUpdate', ...args);
            return true;
        },
        commitUpdate: (node, updatePayload, type, oldProps, newProps) => {
            renderInstance(type, node, newProps);
            console.log('commitUpdate', node, updatePayload, type, oldProps, newProps);
        },
        commitTextUpdate: (node, oldText, newText) => {
            console.log('commitTextUpdate', node, oldText, newText);
        },
        removeChild: (parentNode, childNode) => {
            console.log('removeChild', parentNode, childNode);
            remove(childNode);
        },
        canHydrateInstance: (instance, type, props) => {
            console.log('canHydrateInstance', instance, type, props);
            if (!isReactFigmaNode(instance) || instance.type.toLowerCase() !== type) {
                return null;
            }
            return instance;
        },
        hydrateInstance: (instance, type, props) => {
            console.log('hydrateInstance', instance, type, props);
            return renderInstance(type, instance.type.toLowerCase() === type ? instance : null, props);
        },
        getFirstHydratableChild: parentInstance => {
            console.log('getFirstHydratableChild', parentInstance);
            return getFirstChild(parentInstance);
        },
        getNextHydratableSibling: instance => {
            console.log('getNextHydratableSibling', instance);
            return getNextChildren(instance);
        },
        didNotHydrateContainerInstance: (...args) => {
            console.log('didNotHydrateContainerInstance', ...args);
        },
        didNotFindHydratableContainerInstance: (...args) => {
            console.log('didNotFindHydratableContainerInstance', ...args);
        },
        didNotFindHydratableInstance: (...args) => {
            console.log('didNotFindHydratableInstance', ...args);
        },
        didNotFindHydratableTextInstance: (parentType, parentProps, parentInstance, text) => {
            console.log('didNotFindHydratableTextInstance', parentType, parentProps, parentInstance, text);
        },
        didNotHydrateInstance: (parentType, parentProps, parentInstance, instance) => {
            console.log('didNotHydrateInstance', parentType, parentProps, parentInstance, instance);
        },
        commitMount: (instance, type, newProps, internalInstanceHandle) => {
            console.log('commitMount', instance, type, newProps, internalInstanceHandle);

            if (type === 'page') {
                groupsProcessor.mountGroups();
            }
        }
    };

    const reconciler = createReconciler(HostConfig);

    const container = reconciler.createContainer(rootNode, true, true);
    const lastPage = figma.currentPage;
    const tempPage = figma.createPage();
    figma.currentPage = tempPage;
    reconciler.updateContainer(jsx, container);
    figma.currentPage = lastPage;
    tempPage.remove();
};
