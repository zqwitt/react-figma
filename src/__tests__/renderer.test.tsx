import * as React from 'react';
import { render } from '../renderer';
import { Rectangle, Page, Text, Group } from '..';
import { createFigma } from 'figma-api-stub';

describe('renderer', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('render single component', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        render(<Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />, figma.currentPage);
        expect(figma.createRectangle).toHaveBeenCalledTimes(1);
        expect(figma.root).toMatchSnapshot();
    });

    it('rerender single component', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        render(<Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />, figma.currentPage);
        render(<Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />, figma.currentPage);
        expect(figma.createRectangle).toHaveBeenCalledTimes(1);
        expect(figma.root).toMatchSnapshot();
    });

    it('insert new component between', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Text characters="test" />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(2);
        expect(figma.createText).toHaveBeenCalledTimes(1);
        expect(figma.createPage).toHaveBeenCalledTimes(3);
        expect(figma.root).toMatchSnapshot();
    });

    it('remove component between', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Text characters="test" />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(2);
        expect(figma.createText).toHaveBeenCalledTimes(1);
        expect(figma.createPage).toHaveBeenCalledTimes(3);
        expect(figma.root).toMatchSnapshot();
    });

    it('remove component between (equal components)', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#0048ff' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#00ff00' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#0048ff' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(3);
        expect(figma.createPage).toHaveBeenCalledTimes(3);
        expect(figma.root).toMatchSnapshot();
    });

    it('text instance without Text component', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} /> fff
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(2);
        expect(figma.createText).toHaveBeenCalledTimes(0);
        expect(figma.root).toMatchSnapshot();
    });

    it('text instance without Text component (with hydration page)', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        render(<Page></Page>, figma.root);
        render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} /> fff
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>,
            figma.root
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(2);
        expect(figma.createText).toHaveBeenCalledTimes(0);
        expect(figma.root).toMatchSnapshot();
    });

    it('creates single group', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        figma.group = jest.fn().mockImplementation(figma.group);

        render(
            <Page>
                <Group>
                    <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                </Group>
            </Page>,
            figma.root
        );

        expect(figma.group).toHaveBeenCalledTimes(1);
        expect(figma.root).toMatchSnapshot();
    });

    it('creates nested groups', () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        figma.group = jest.fn().mockImplementation(figma.group);

        render(
            <Page>
                <Group>
                    <Group>
                        <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                    </Group>
                    <Group>
                        <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                        <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                    </Group>
                </Group>
            </Page>,
            figma.root
        );

        expect(figma.group).toHaveBeenCalledTimes(3);
        expect(figma.root).toMatchSnapshot();
    });
});
