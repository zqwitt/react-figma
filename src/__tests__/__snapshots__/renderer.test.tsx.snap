// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer creates nested groups 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        ChildrenMixinStub {
          "children": Array [
            ChildrenMixinStub {
              "children": Array [
                LayoutMixinStub {
                  "fills": Array [
                    Object {
                      "color": Object {
                        "b": 0,
                        "g": 1,
                        "r": 0.07058823529411765,
                      },
                      "type": "SOLID",
                    },
                  ],
                  "height": 100,
                  "parent": [Circular],
                  "pluginData": Object {
                    "isReactFigmaNode": "true",
                    "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
                  },
                  "type": "RECTANGLE",
                  "width": 200,
                },
                LayoutMixinStub {
                  "fills": Array [
                    Object {
                      "color": Object {
                        "b": 0,
                        "g": 1,
                        "r": 0.07058823529411765,
                      },
                      "type": "SOLID",
                    },
                  ],
                  "height": 100,
                  "parent": [Circular],
                  "pluginData": Object {
                    "isReactFigmaNode": "true",
                    "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
                  },
                  "type": "RECTANGLE",
                  "width": 200,
                },
              ],
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
              },
              "type": "GROUP",
            },
            ChildrenMixinStub {
              "children": Array [
                LayoutMixinStub {
                  "fills": Array [
                    Object {
                      "color": Object {
                        "b": 0,
                        "g": 1,
                        "r": 0.07058823529411765,
                      },
                      "type": "SOLID",
                    },
                  ],
                  "height": 100,
                  "parent": [Circular],
                  "pluginData": Object {
                    "isReactFigmaNode": "true",
                    "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
                  },
                  "type": "RECTANGLE",
                  "width": 200,
                },
              ],
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
              },
              "type": "GROUP",
            },
          ],
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
          },
          "type": "GROUP",
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer creates single group 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        ChildrenMixinStub {
          "children": Array [
            LayoutMixinStub {
              "fills": Array [
                Object {
                  "color": Object {
                    "b": 0,
                    "g": 1,
                    "r": 0.07058823529411765,
                  },
                  "type": "SOLID",
                },
              ],
              "height": 100,
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
                "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
              },
              "type": "RECTANGLE",
              "width": 200,
            },
          ],
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
          },
          "type": "GROUP",
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer insert new component between 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "characters": "test",
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
          },
          "type": "TEXT",
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer remove component between (equal components) 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 1,
                "g": 0.2823529411764706,
                "r": 0,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#0048ff\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer remove component between 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer render single component 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer rerender single component 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer text instance without Text component (with hydration page) 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer text instance without Text component 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;
